eogrow/pipelines/split_grid.py

```diff
@@ -0,0 +1,184 @@
+"""Implements a pipeline that creates a finer grid and splits EOPatches accordingly."""
+import logging
+from collections import defaultdict
+from typing import Dict, List, Literal, Tuple, Union
+
+import fs
+import geopandas as gpd
+from pydantic import Field
+
+from eolearn.core import EONode, EOWorkflow, FeatureType, LoadTask, OverwritePermission, SaveTask
+from sentinelhub import BBox
+
+from ..core.area.batch import BatchAreaManager
+from ..core.area.utm import UtmZoneAreaManager
+from ..core.pipeline import Pipeline
+from ..tasks.spatial import SpatialSliceTask
+from ..utils.fs import LocalFile
+from ..utils.grid import split_bbox
+from ..utils.types import Feature, FeatureSpec
+
+LOGGER = logging.getLogger(__name__)
+
+NamedBBox = Tuple[str, BBox]
+
+
+class SplitGridPipeline(Pipeline):
+    """Pipeline that creates a finer grid and splits EOPatches accordingly.
+
+    The new grid is output as a geopackage file, which can be used with a `CustomAreaManager`.
+    """
+
+    class Schema(Pipeline.Schema):
+        input_folder_key: str = Field(
+            description="A storage manager key pointing to the folder from where data will be loaded."
+        )
+        eopatch_output_folder_key: str = Field(
+            description="A storage manager key pointing to the folder to where data will be saved."
+        )
+        grid_output_folder_key: str = Field(
+            description="A storage manager key of where to save the resulting split grid."
+        )
+        subsplit_grid_filename: str = Field(
+            description="Filename of new grid, which can be used in `CustomAreaManager`.", regex=r"^.+\.gpkg$"
+        )
+        features: List[Feature] = Field(description="Which features will be loaded and joined.")
+        raise_misaligned: bool = Field(
+            True,
+            description=(
+                "If True this will raise an error if splitting or joining any spatial raster EOPatch feature would "
+                "cause a misalignment. If False, misalignment issues will be ignored."
+            ),
+        )
+
+        split_x: int = Field(2, description="Number of splits done on each EOPatch along the x-axis.")
+        split_y: int = Field(2, description="Number of splits done on each EOPatch along the y-axis.")
+
+        buffer: Union[Literal["auto"], Tuple[float, float]] = Field(
+            "auto",
+            description=(
+                "How large is the original EOPatch buffer, provided as `(buffer_x, buffer_y)` in CRS units. Applies"
+                " same buffer to split EOPatches. The pipeline tries to obtain this information from area manager, but"
+                " not all area manager classes are supported."
+            ),
+        )
+        # TODO: could add AOI file?
```

Comment:
Please explain this, not sure I understand how this would work.

Reply:
If an AOI file is provided, all EOPatches that do not lie inside the area are discarded.
This was done by previous implementation, since Batch and UTM area managers carry an AOI.

Reply:
Interesting use case, one could use this pipeline without splitting further and just using the input AOI as a filter, e.g. for running several per-region experiments.

But I don't see how this functionality would be useful when doing a split, not how the split functionality would be useful when doing the filtering, so IDK.

Reply:
Let's see if i can ASCII-art myself out of this...

Assume this is the AOI and is covered by a single EOPatch in the original grid.
```
┌──────┐
│ ▉    │
│ ▉▉   │
│ ▉▉▉▉ │
│  ▉▉  │
└──────┘
```
By splitting into 2x2 we currently get 
```
┌───┐┌───┐
│ ▉ ││   │
│ ▉▉││   │
└───┘└───┘
┌───┐┌───┐
│ ▉▉││▉▉ │
│  ▉││▉  │
└───┘└───┘
```
but there is no need for the upper right EOPatch. However in order for the pipeline to know that, it would need to have access to the AOI. One option is to take the one from the AreaManager, but it'd say its better it's an explicit input.

Reply:
I think it's good if such an input functionality is added, but would set the default to the AOI from AreaManager

Reply:
nice art!

Reply:
What if the default is that no such filtration is used?

Reply:
I assumed that _original AOI_ == _no filtration_, or did you mean that also the upper right split eopatch is retained in this case?

Reply:
If no AOI is provided, all the resulting EOPatches would be retained. If an AOI is provided, only EOPatches intersecting it would be retained.

Reply:
From the point of view of not making decisions on behalf of the user, I completely agree with this.

However, from the other point of view, we might be pushing people to use unnecessary resources for processing split eopatches which fall out of the AOI they're originally interested in.

idk.. should we vote? I would still vote for the latter, because if the user would provide the finer grid, these extra eopatches wouldn't be present.

Reply:
This can be resolved, right? Due to https://github.com/sentinel-hub/eo-grow/pull/150